Spring Boot AI 서비스 연동 502 에러 해결

23MinL opened the pull request:
## ⭐️ Issue Number
- #27 

## 🚩 Summary

- AI 서비스 아키텍처를 단일 EC2에서 분리된 EC2 구조로 변경하여 VPC 내부 통신 구현
- FastAPI 전용 EC2 인스턴스(Private IP: 10.0.5.114) 생성으로 AI 워크로드 분리
- Spring Boot 설정에서 AI 서비스 URL을 VPC 내부 통신용으로 업데이트

## 🛠️ Technical Concerns
아키텍처 변경

- 기존: Spring Boot EC2 내에서 AI 서비스 동시 실행
- 변경: Spring Boot EC2 ↔ FastAPI 전용 EC2 VPC 내부 통신
- 장점: 독립적 확장, 리소스 격리, 장애 전파 방지

네트워크 설정

- VPC 내부 통신 (10.0.10.157 ↔ 10.0.5.114)
- 보안 그룹 최적화로 8000 포트 VPC 내부만 접근 허용
- Network ACL 설정으로 네트워크 레벨 보안 강화

성능 최적화

- Connection Pool 설정으로 연결 재사용 효율성 증대
- Circuit Breaker 패턴으로 장애 시 빠른 복구
- AI 추천 결과 캐싱(30분 TTL)으로 응답 속도 향상

## 🙂 To Reviewer
테스트 확인 사항
1. **VPC 내부 통신 테스트**:
   ```bash
   # Spring Boot EC2에서 FastAPI 연결 확인
   curl -f http://10.0.5.114:8000/health
   ```

2. **AI 추천 API 호출 테스트**:
   ```bash
   curl -X POST http://api.ururu.shop/api/ai/groupbuy-recommendations/by-profile \
   -H "Content-Type: application/json" \
   -d '{"topK": 40}'
   ```

3. **헬스체크 모니터링**:
   - Spring Boot Actuator: `/actuator/health`
   - FastAPI Health: `/health`

주요 변경사항 코드 리뷰 포인트
- `application-prod.yml`: AI 서비스 URL 및 연결 설정
- `AiServiceClient.java`: VPC 통신 최적화 및 예외 처리
- FastAPI 서비스: Two-Tower 추천 시스템 구현

## 📋 To Do
- [x] FastAPI EC2 인스턴스 생성 및 보안 그룹 설정
- [x] Spring Boot AI 서비스 설정 VPC 내부 통신용으로 변경
- [x] AI 서비스 클라이언트 안정성 개선
- [x] FastAPI Two-Tower 추천 시스템 초기 구현
- [ ] 실제 상품 데이터 임베딩 생성 및 Faiss 인덱스 구축
- [ ] AI 추천 시스템 성능 테스트 및 튜닝
- [ ] 프론트엔드 AI 추천 기능 연동 테스트

### 배포 확인 사항
- [x] Spring Boot EC2에서 기존 AI 서비스 정리 완료 (16.43GB 용량 회수)
- [x] FastAPI EC2 CI/CD 파이프라인 구축 완료
- [ ] 운영 환경 배포 후 AI 추천 API 정상 동작 확인 필요

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


**Chores**
  * CI/CD 및 Docker 관련 워크플로우가 프로덕션 환경 중심으로 재구성되었습니다.
  * 개발 환경용 docker-compose 파일 및 퍼포먼스 테스트 워크플로우가 삭제되었습니다.
  * 환경 변수 관리 및 설정 파일 경로가 통일되고, 배포 실패 시 GitHub 이슈 자동 생성 기능이 추가되었습니다.
  * Dockerfile 및 컴포즈 파일에서 로그/캐시 디렉터리 관리와 네트워크 구성이 개선되었습니다.

**New Features**
  * 프로덕션 AI 서비스용 Docker Compose 파일이 신규 추가되었습니다.
  * 설정에 Redis 및 Spring Boot 연동 관련 옵션이 추가되었습니다.

**Bug Fixes**
  * 환경 변수 및 토큰 사용 방식이 통일되고, VPC 환경에 맞게 내부 통신 주소가 수정되었습니다.

**Documentation**
  * 배포 및 헬스체크 보고 메시지가 개선되고, 자세한 배포 안내가 추가되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

23MinL committed:
fix: Spring Boot AI 서비스 연동 502 에러 해결

- AI 서비스 URL 설정 수정
- nginx 프록시 설정 추가
- 헬스체크 엔드포인트 연동 확인

23MinL committed:
refactor: AI 서비스를 별도 EC2로 분리하여 VPC 내부 통신 구현

- FastAPI AI 서비스 전용 EC2 인스í
- VPC 내부 통신을 위한 Pri5.114) 활용
- Spring Boot ↔ FastAPI 간 네트워크 분리로 성능 및 확장성 향상
- AI 워크로드와 백엔드 워크로드 독립적 관리 가능

23MinL committed:
fix: AI 서비스 배포 환경 설정 완료

- GHCR_TOKEN 기반 인증으로 통일
- VPC 내부 통신 환경변수 설정 완료
- Docker 헬스체크 경로 및 토큰 수정
- 한국어 특화 AI 모델 운영 환경 최적화

23MinL committed:
fix: Docker 헬스체크 워크플로우 토큰 및 경로 수정

1000hyehyang reviewed (approved):
일단 머지 후 배포 과정에서 cicd 문제와 연동 문제가 생기지 않는지 확인해봅시다! 고생 많으셨습니다

23MinL committed:
fix: AI 서비스 배포 환경 설정

taehyun32 reviewed (commented):
DM으로 전달했습니다

23MinL committed:
refactor: Docker Compose 파일명을 AI 특화로 변경

- docker-compose.yml → docker-compose-ai.yml
- docker-compose-prod.yml → docker-compose-ai-prod.yml
- Spring Boot 리포지토리와 네이밍 충돌 방지
- CI/CD 파이프라인 경로 수정

23MinL committed:
refactor: CI/CD 및 Docker 구성 최적화

23MinL commented on .github/workflows/ci-cd.yml:
```diff
@@ -187,25 +138,129 @@ jobs:
+            SPRING_BOOT_BASE_URL=http://${{ secrets.SPRING_BOOT_PRIVATE_IP }}:8080
+            BUILD_TARGET=production
+            EOF
+            
```
SPRING_BOOT_PRIVATE_IP secrets 추가 완료 + 환경변수 중복 문제 수정 완료

23MinL committed:
fix: EC2 배포 시 작업 디렉토리 자동 생성

23MinL commented on .github/workflows/ci-cd.yml:
```diff
@@ -187,25 +138,129 @@ jobs:
+          key: ${{ secrets.AI_EC2_SSH_KEY }}        # 새로운 EC2 키
+          port: 22
+          timeout: 600s
+          script: |
+            set -e
+            echo "🚀 FastAPI AI 서비스 배포 시작"
+            
+            # 기존 컨테이너 상태 확인
+            if docker ps | grep ururu-ai-service; then
+              echo "📊 기존 AI 서비스 로그 확인"
+              docker logs --tail 10 ururu-ai-service
+            fi
+            
+            # 코드 업데이트 (올바른 디렉토리)
+            cd /home/ec2-user/ururu-ai
+            if [ ! -d ".git" ]; then
+              echo "📥 레포지토리 초기 클론"
+              git clone https://github.com/UruruLab/Ururu-AI.git .
+            else
+              echo "🔄 코드 업데이트"
+              git fetch origin
+              git checkout main
+              git reset --hard origin/main
+            fi
```
# Before (문제)
cd /home/ec2-user/ururu-ai  # 디렉토리가 없으면 실패

# After (수정)
mkdir -p /home/ec2-user/ururu-ai  # 디렉토리 자동 생성
cd /home/ec2-user/ururu-ai        # 안전하게 이동

23MinL committed:
fix: Docker 헬스체크 설정 파일 경로 수정

- docker compose config에서 파일명 명시적 지정
- docker-compose-ai.yml 파일 사용하도록 수정
- 'no configuration file provided' 에러 해결

23MinL commented on docker/docker-compose-ai.yml:
```diff
@@ -6,7 +6,7 @@ services:
       - ./config:/config
       - ./scripts:/scripts:ro
     environment:
-      - GITHUB_TOKEN=${GITHUB_TOKEN}
+      - GITHUB_TOKEN=${GHCR_TOKEN}
       - CONFIG_REPO_URL=https://github.com/UruruLab/Ururu-AI-Config.git
       - ENVIRONMENT=${ENVIRONMENT:-production}
```
지금 Ururu AI에서는 PRIVATE_REPO_TOKEN가 아닌 GHCR_TOKEN로 권한을 줘서 사용하고 있으니 무방합니다.

23MinL commented:
<img width="773" height="538" alt="image" src="https://github.com/user-attachments/assets/0a58d0eb-b426-47aa-af34-af6229d985ce" />
스킵된 부분 외에 정상작동 하는 것을 고려해서 cicd 및 health check는 main에서만 구동되게 커밋하겠습니다.

23MinL committed:
fix : cicd 및 healthcheck main 브랜치에서 구동 수정

23MinL commented on .github/workflows/ci-cd.yml:
```diff
@@ -16,8 +16,8 @@ jobs:
     env:
-      ENVIRONMENT: development
-      DEBUG: true
+      ENVIRONMENT: production
+      DEBUG: false
       LOG_LEVEL: INFO
       GITHUB_ACTIONS: true
       
```
현재 ENVIRONMENT=production으로 설정되어 있는데, 이는 매우 위험합니다.
테스트에서 production 환경변수를 사용하면 실제 운영 DB나 서비스에 접근할 위험이 있습니다.
이전 변경점으로 인해 완전히 제거하는 것이 맞습니다.

23MinL commented on .github/workflows/ci-cd.yml:
```diff
@@ -187,25 +138,130 @@ jobs:
-          echo "- Docker Image: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:latest"
-          echo "- Config File: .env.production (fetched from Config repository)"
-          echo "Actual EC2 deployment will be handled by separate process"
+      - name: Deploy to FastAPI Production EC2
+        uses: appleboy/ssh-action@v0.1.6
+        with:
+          host: ${{ secrets.AI_EC2_HOST }}          # 새로운 FastAPI EC2
+          username: ${{ secrets.AI_EC2_USER }}      # ec2-user
+          key: ${{ secrets.AI_EC2_SSH_KEY }}        # 새로운 EC2 키
+          port: 22
+          timeout: 600s
+          script: |
```
SSH 배포 스크립트가 더 간결하고 명확해졌습니다.
.env.production 파일을 직접 업로드하는 방식이 더 안전하고 효율적입니다.

이 구조로 적용하겠습니다.

23MinL commented on .github/workflows/docker-health-check.yml:
```diff
@@ -89,14 +95,24 @@ jobs:
+          # 환경별 필수 변수 체크
+          echo "🔍 Checking required environment variables"
+          cd docker
+          if grep -q "SPRING_BOOT_BASE_URL" config/.env.production; then
+            echo "✅ SPRING_BOOT_BASE_URL configured"
+          else
+            echo "⚠️ SPRING_BOOT_BASE_URL not found in production config"
+          fi
```
이 부분은 유지하는 것이 좋겠습니다!

SPRING_BOOT_BASE_URL이 없으면 FastAPI에서 Spring Boot에 연결할 수 없어 서비스 장애가 발생할 수 있습니다.
배포 전에 필수 환경변수를 검증하는 것은 안전장치 역할을 합니다.

23MinL commented on .github/workflows/docker-health-check.yml:
```diff
@@ -26,61 +26,67 @@ jobs:
+            echo "✅ .env.development file exists in docker context"
           else
-            echo "❌ .env.development file missing"
+            echo "❌ .env.development file missing in docker context"
             exit 1
           fi
-          if [ -f ".env.production" ]; then
-            echo "✅ .env.production file exists"
+          
+          if [ -f "./docker/config/.env.production" ]; then
+            echo "✅ .env.production file exists in docker context"
           else
-            echo "❌ .env.production file missing"
+            echo "❌ .env.production file missing in docker context"
             exit 1
           fi
```
production 배포에서는 .env.development 파일이 필요 없으므로 해당 검증을 제거하는 것이 맞습니다.
.env.production 파일 확인만 유지하는 것이 적절합니다.

23MinL committed:
refactor: CI/CD 파이프라인에서 test job 제거

- test job 완전 제거로 DB 접속 위험성 차단
- build-and-push job의 test 의존성 제거
- 파이프라인 단순화로 배포 속도 향상

23MinL committed:
fix: SSH 배포에서 GITHUB_SHA 환경변수 전달 개선

23MinL commented on .github/workflows/ci-cd.yml:
```diff
@@ -187,25 +138,130 @@ jobs:
+              curl -X POST "http://localhost:8000/api/vector/embeddings/batch?batch_size=100&force_recreate=false" || echo "임베딩 재생성 요청 실패"
+            fi
+            
+            echo "🎉 FastAPI AI 서비스 배포 완료"
+            echo "$(date): FastAPI AI 서비스 배포 완료 - commit: $GITHUB_SHA" >> /home/ec2-user/deployment.log
```
env 섹션 추가: GITHUB_SHA: ${{ github.sha }} 환경변수 설정
envs 속성 추가: envs: GITHUB_SHA로 SSH 스크립트에 환경변수 전달
주석 정리: 불필요한 주석 제거로 코드 정리
